### PR TITLE
[Feature:HelpQueue] Sort button for Office Hours Queue

### LINKED ISSUE #9789: Sort/new statistics of office hours queue history
**What problem are you trying to solve with Submitty**
Sometimes we have to take screenshots of who we helped and currently there is no way to only take screenshots of who I helped. We can only send screenshots of a collection of people who were helped during the time period including me.

**Describe the way you'd like to solve this problem**
1.  Sort by "help started by" and other columns
2. A search for a specific name/id tag by "help started by"
3. Have a separate statistic of who you helped in that day
4. Do all of the above?

**Additional context**
image
Currently no way to filter through queue history

### PR DESCRIPTION
**Pull request by ArnavPawar**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [ ] Screenshots are attached to GitHub PR if visual/UI changes were made

### What is the current behavior?

Closes #9789  Sorts by the reviewer in the office hours history so that the reviewer can easily see how many students they helped

### What is the new behavior?
When the button is pressed it should sort by 'Help Started By'
    - has a toggle behavior to be up, down, and off
image
image
image

### Other information?

Test this feature it by:
1) running this Python script to populate the office hours queue:
python3 GIT_CHECKOUT/Submitty/.setup/bin/populate_office_hours_queue.py

2) Pressing on the blue column labeled 'Helped Started By' to test the feature

### CODE REVIEW

**Review by DarthNyan — CHANGES_REQUESTED** [review]
It seems this is still WIP so it should be marked as a draft. Also, be sure to follow the title guidelines found [here](https://submitty.org/developer/getting_started/make_a_pull_request).

This seems like a cool feature. I can't wait to see it complete.

**Comment by williamjallen** [comment]
@ArnavPawar I see that you have requested a review from someone.  If you are finished making changes and are ready to open this up for review, you should mark it as such by removing the "draft" status.

File: site/app/templates/officeHoursQueue/QueueHistory.twig [hunk 1]
```diff
@@ -84,3 +92,37 @@
     </tbody>
   </table>
 </div>
+
+<script>
```
Comment (williamjallen) [inline site/app/templates/officeHoursQueue/QueueHistory.twig:96]: It would be better to move this to a dedicated JS file like `site/public/js/office-hours-queue.js`.

File: site/app/libraries/database/DatabaseQueries.php [hunk 2]
```diff
@@ -7275,8 +7275,20 @@ public function getCurrentQueue() {
         $this->course_db->query($query, [$this->core->getDateTimeNow()->format('Y-m-d')]);
         return $this->course_db->rows();
     }
-
     public function getPastQueue() {
+        $off  = "ORDER BY ROW_NUMBER";
+        $up   = "ORDER BY helper.user_id";
+        $down = "ORDER BY helper.user_id DESC";
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:7281]: You could move the "ORDER BY" part to the main body of the SQL query rather than duplicating it here.

File: site/app/libraries/database/DatabaseQueries.php [hunk 3]
```diff
@@ -7275,8 +7275,20 @@ public function getCurrentQueue() {
...
+        $orderType = "";
+        if ($_COOKIE['sortStateTimeEntered'] == "up") {
+            $orderType = $up;
+        }
+        elseif ($_COOKIE['sortStateTimeEntered'] == "down") {
+            $orderType = $down;
+        }
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:7288]: Is this something which should be saved in the browser via cookies and determined on the server side like is being done here, or something we should do on the client side instead?

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/templates/officeHoursQueue/QueueHistory.twig [hunk 4]
```diff
@@ -84,3 +92,37 @@
...
+            } else if (Cookies.get('sortStateTimeEntered') == "up") {
+                Cookies.set('sortStateTimeEntered', 'down'); // Save to cookie
+                sortIndicator.innerHTML = '<i class="fas fa-sort-down"></i>';
+            } else if (Cookies.get('sortStateTimeEntered') == "down") {
+                Cookies.set('sortStateTimeEntered', 'off'); // Save to cookie
+                sortIndicator.innerText = ""; // Reset to down arrow for off state
+                console.log(Cookies.get('sortStateTimeEntered'))
+            }
+        }
+        window.location.reload();
```
Comment (DarthNyan) [inline site/app/templates/officeHoursQueue/QueueHistory.twig:126]: I didn't like how the page would reload and send me up to the top every time I clicked the sort button. I think avoiding refreshing the page and just moving the table entries around would feel better to users.

File: site/app/libraries/database/DatabaseQueries.php [hunk 5]
```diff
@@ -7275,8 +7275,20 @@ public function getCurrentQueue() {
         $this->course_db->query($query, [$this->core->getDateTimeNow()->format('Y-m-d')]);
         return $this->course_db->rows();
     }
-
     public function getPastQueue() {
+        $off  = "ORDER BY ROW_NUMBER";
+        $up   = "ORDER BY helper.user_id";
+        $down = "ORDER BY helper.user_id DESC";
+        $orderType = "";
+        if ($_COOKIE['sortStateTimeEntered'] == "up") {
```
Comment (DarthNyan) [inline site/app/libraries/database/DatabaseQueries.php:7283]: I'm getting a warning about 'sortStateTimeEntered' being an undefined array key when I access the queue page. I second William's question about sorting on the client side and not using cookies.

**Review by DarthNyan — CHANGES_REQUESTED** [review]

File: site/public/js/office-hours-queue.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        else if (localStorage.getItem('sortIndicator') === 'down') {
+            localStorage.setItem('sortIndicator', 'off');
+            sortIndicator.innerHTML = '<i class="fa-solid fa-sort"></i>';
+        }
+    }
+    adjustRows();
+}
+/* eslint prefer-arrow-callback: [ "error", { "allowNamedFunctions": true } ] */
+function adjustRows() {
+    const sortIndicator = document.getElementById('sortIndicator');
```
Comment (DarthNyan) [inline site/public/js/office-hours-queue.js:46]: you can delete this variable since it's not used.

File: site/public/js/office-hours-queue.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    const rows = $('.queue_history_row').toArray();
+    rows.sort(function (a, b) {
```
Comment (DarthNyan) [inline site/public/js/office-hours-queue.js:48]:
```suggestion
    rows.sort((a, b) => {
```

File: site/public/js/office-hours-queue.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
+//this helps update the frontend when the page refreshes because without this the sort icon would reset and the sort state would not
+/* eslint prefer-arrow-callback: [ "error", { "allowNamedFunctions": true } ] */
+document.addEventListener('DOMContentLoaded', function() {
```
Comment (DarthNyan) [inline site/public/js/office-hours-queue.js:3]:
```suggestion
document.addEventListener('DOMContentLoaded', () => {
```

File: site/app/libraries/database/DatabaseQueries.php [hunk 9]
```diff
@@ -7323,14 +7321,13 @@ public function getPastQueue() {
             AS h1
             ON queue.user_id = h1.uid AND queue.queue_code = h1.qc
             WHERE time_in > ? AND current_state IN ('done')
-            ORDER BY row_number
+            ORDER BY ROW_NUMBER
```
Comment (DarthNyan) [inline site/app/libraries/database/DatabaseQueries.php:7324]: Do you still need to change the query now that you're sorting in javascript?

**Review by DarthNyan — CHANGES_REQUESTED** [review]
I got a string of emails saying you'd committed a bunch of lint fixes. I left a few comments that should help fix the remaining errors.

**Review by RitaLei123 — APPROVED** [review]
The three step review process:
**Is the proposed change a good idea?**
Yes it is a good idea as it will allow an easier view of the queue.

**Did the author take the right approach?**
Yes, the author adjusted the relevant files. I looked through the code but not sure how everything works but it looks correct.

**Are all the implementation details correct?**
Yes the implementation details are correct and it works. 

**Manual testing:**
Reproduce on main: The function was not there
Check topic branch
Reproduce on main again: It works very smoothly

**A few other things to check:**
If a UI change: light/dark modes, different screen sizes: It works with different screen sizes.

File: site/app/templates/officeHoursQueue/QueueHistory.twig [hunk 10]
```diff
@@ -22,20 +22,28 @@
...
+
         <th scope="col" class="mobile-hide">Removed By</th>
+
         {% if viewer.isGrader() %}
           <th scope="col">Restore</th>
         {% endif %}
       </tr>
     </thead>
-    <tbody>
+    <tbody id="queueHistoryTable">
```
Comment (DarthNyan) [inline site/app/templates/officeHoursQueue/QueueHistory.twig:39]: Same here

File: site/app/templates/officeHoursQueue/QueueHistory.twig [hunk 11]
```diff
@@ -22,20 +22,28 @@
...
       {% set history_count = 0 %}
       {% for entry in viewer.getPastQueue() %}
+
         {% if (history_count < 25 or viewer.getFullHistory()) and (viewer.isGrader() or entry['user_id'] == viewer.getUserId()) %}
           {% set history_count = history_count + 1 %}
           <tr class="queue_history_{{entry['queue_code']}} queue_history_row">
-            <th scope="row">{{history_count}}</th>
+            <th scope="row" class="numberCount">{{history_count}}</th>
```
Comment (DarthNyan) [inline site/app/templates/officeHoursQueue/QueueHistory.twig:46]: Same here

File: site/app/templates/officeHoursQueue/QueueHistory.twig [hunk 12]
```diff
@@ -61,7 +69,7 @@
 
             <td>{{ entry['helped_today'] }}</td>
             <td>{{ entry['times_helped'] }}</td>
-            <td class="mobile-hide">{{ entry['help_started_by'] }}</td> 
+            <td class="mobile-hide helpStarted">{{ entry['help_started_by'] }}</td>
```
Comment (DarthNyan) [inline site/app/templates/officeHoursQueue/QueueHistory.twig:72]: Same here

**Review by DarthNyan — CHANGES_REQUESTED** [review]
Just a few style things. Other than that, things look good to me.

**Review by DarthNyan — APPROVED** [review]
Looks good to me!

**Review by benjaminlee755 — APPROVED** [review]
Everything looks good to me. The history gets sorted by Help Started

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1]: This file only has trivial changes which are unrelated to the rest of the PR, so the changes should be reverted to prevent unnecessary conflicts with other PRs.

File: site/public/js/office-hours-queue.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    const sortIndicator = document.getElementById('sort-indicator');
+    let sortState = localStorage.getItem('sort-indicator');
+
+    if (sortState === null) {
+        sortState = 'off';
+        localStorage.setItem('sort-indicator', sortState);
+    }
+
+    if (sortState === 'up') {
+        sortIndicator.innerHTML = '<i class="fas fa-sort-up"></i>';
```
Comment (williamjallen) [inline site/public/js/office-hours-queue.js:13]: Using `.innerHTML` opens the possibility for XSS attacks to pop up in the future.  You could just use jQuery to change the class instead rather than adding some HTML.  Apply throughout.

File: site/public/js/office-hours-queue.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+function adjustRows() {
+    const rows = $('.queue_history_row').toArray();
+    rows.sort((a, b) => {
+        if (localStorage.getItem('sort-indicator') === 'up') {
+            if ($(a).find('.help-started').text()>$(b).find('.help-started').text()) {
+                return -1;
+            }
+            else {
+                return 1;
+            }
```
Comment (williamjallen) [inline site/public/js/office-hours-queue.js:52]: You could simplify this significantly with a ternary expression.  Apply throughout.

```suggestion
            return $(a).find('.help-started').text() > $(b).find('.help-started').text() ? -1 : 1;
```

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/public/js/office-hours-queue.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
+//this helps update the frontend when the page refreshes because without this the sort icon would reset and the sort state would not
+/* eslint prefer-arrow-callback: [ 'error', { 'allowNamedFunctions': true } ] */
```
Comment (williamjallen) [inline site/public/js/office-hours-queue.js:2]: Why is this eslint command here?

File: site/public/js/office-hours-queue.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    else if (sortState === 'down') {
+        sortIndicator.innerHTML = '<i class="fas fa-sort-down"></i>';
+    }
+    adjustRows();
+});
+
+// eslint-disable-next-line no-unused-vars
+/* eslint prefer-arrow-callback: [ "error", { "allowNamedFunctions": true } ] */
+function toggleSort(column) {
```
[the remark this answers is not in the document]
Comment (williamjallen) [reply site/public/js/office-hours-queue.js:26]: @DarthNyan is correct here.  This is a good suggestion.

**Review by williamjallen — CHANGES_REQUESTED** [review]
My only remaining comments have to do with minor eslint issues.  Otherwise, the code looks good.  I did not test the functionality.

File: site/public/js/office-hours-queue.js [hunk 17]
```diff
@@ -0,0 +1,61 @@
+//this helps update the frontend when the page refreshes because without this the sort icon would reset and the sort state would not
+document.addEventListener('DOMContentLoaded', () => {
+    const sortIndicator = document.getElementById('sort-indicator');
+    let sortState = localStorage.getItem('sort-indicator');
+
+    if (sortState === null) {
+        sortState = 'off';
+        localStorage.setItem('sort-indicator', sortState);
```
Comment (cjreed121) [inline site/public/js/office-hours-queue.js:8]: This will set the key for the site globally as `sort-indicator`. Can you be more specific and add office hours to it or something?

File: site/app/templates/officeHoursQueue/QueueHistory.twig [hunk 18]
```diff
@@ -83,4 +92,4 @@
       {% endfor %}
     </tbody>
   </table>
-</div>
+</div>
```
Comment (cjreed121) [inline site/app/templates/officeHoursQueue/QueueHistory.twig:95]: Unnecessary change

File: site/public/js/office-hours-queue.js [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    else if (sortState === 'off') {
+        sortIndicator.classList.add('fa-solid', 'fa-sort');
+    }
+    else if (sortState === 'down') {
+        sortIndicator.classList.add('fas', 'fa-sort-down');
+    }
+
+    adjustRows();
+});
+/* exported toggleSort */
```
Comment (cjreed121) [inline site/public/js/office-hours-queue.js:23]: We generally put this at the very top of the file

**Review by cjreed121 — CHANGES_REQUESTED** [review]

**Review by cjreed121 — APPROVED** [review]

**Review by VVVITH — APPROVED** [review]
